### PR TITLE
wgsl: Fix conditions on break from continuing

### PR DESCRIPTION
**Pull request by dneto0**
A break from a continuing clause must be conditional, where that
condition does nothing else, and that conditional must come last
in the continuing clause.

- If it's unconditional then there is no actual backedge for the loop,
  and so it's not a loop in the first place.
- if it comes too early then it breaks the rule that the backedge
  must post-dominate the head of the continuing clause.
- Still allows conditional exit from the loop at the bottom of the loop.

### CODE REVIEW

File: wgsl/index.bs [hunk 1]
```diff
@@ -1229,7 +1229,13 @@ after the body of the nearest-enclosing [[#loop-statement]]
 or [[#switch-statement]].
 
 When a `break` statement is placed such that it would exit from a loop's [[#continuing-statement]],
-the break statement must appear last in that [[#continuing-statement]].
+then:
+
+* The `break` statement must appear as either:
+    * The only statement in the true-branch clause of an `if` that has no `else` clause and no `elseif` clauses, or
```
Comment (dneto0) [inline wgsl/index.bs:1235]: Essentially, this says

* must be a break-if or break-unless that appears last in the continuing construct

Except that we just removed break-if and break-unless.

**Comment by dneto0** [comment]
> Oh wow. I wonder if we can just say "no break from continuing" for now?

It's common to exit the loop from its bottom.   We don't want to ban that, and I don't want to ban that in a way that's coupled to whether someone has done a "continue" statement earlier.

**Comment by kvark** [comment]
Since `continuing` block is often the product of the shader pipeline, and not really used by the shader authors (at least until they start using WGSL), would you be able to point me to some source code, say, on [shader playground](http://shader-playground.timjones.io/) that compiles to SPIR-V with a break inside continuing? Still have a hard time understanding this use case.

**Comment by grorg** [comment]
Discussed at the [2020-06-09 meeting](https://docs.google.com/document/d/1ap37K-4SaybRS9Qdt7fj8SNkwOnoH3GyVBPKLYDMm8s/).

**Comment by jdashg** [comment]
I think that we should either:
1) Allow `if (foo) { break; } else {}`
or
2) Forbid `if (foo) {} else { break; }`

**Comment by dneto0** [comment]
> I think that we should either:

Ah, I didn't get your meaning during the meaning.

Yes, I would be happy to allow `if (foo) { break; } else {}`

**Comment by dneto0** [comment]
>  would you be able to point me to some source code, say, on shader playground that compiles to SPIR-V with a break inside continuing? Still have a hard time understanding this use case.

Example: 
http://shader-playground.timjones.io/b110d75b288ceadf381b5535e01ce384

The do-while loop maps directly to exiting from the bottom of a loop.

The source for the example is:

```glsl
#version 450

shared float w;

void main() {
  float f = 52.0;
  float err;
  do {
    const float next = 1 + 1 / err;
    const int bits = floatBitsToInt(next);
    if ((bits & 0x100) == 0) { f = f + 1.0; continue; }
    err = next - f;
    f = next;
  } while(err > 1e-5);
  w = f;
}
```

The SPIR-V emitted (directly by glslang, and no further processing):

```
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint GLCompute %main "main"
               OpExecutionMode %main LocalSize 1 1 1
               OpSource GLSL 450
               OpName %main "main"
               OpName %f "f"
               OpName %next "next"
               OpName %err "err"
               OpName %bits "bits"
               OpName %w "w"
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
      %float = OpTypeFloat 32
%_ptr_Function_float = OpTypePointer Function %float
   %float_52 = OpConstant %float 52
    %float_1 = OpConstant %float 1
        %int = OpTypeInt 32 1
%_ptr_Function_int = OpTypePointer Function %int
    %int_256 = OpConstant %int 256
      %int_0 = OpConstant %int 0
       %bool = OpTypeBool
%float_9_99999975en006 = OpConstant %float 9.99999975e-006
%_ptr_Workgroup_float = OpTypePointer Workgroup %float
          %w = OpVariable %_ptr_Workgroup_float Workgroup
       %main = OpFunction %void None %3
          %5 = OpLabel
          %f = OpVariable %_ptr_Function_float Function
       %next = OpVariable %_ptr_Function_float Function
        %err = OpVariable %_ptr_Function_float Function
       %bits = OpVariable %_ptr_Function_int Function
               OpStore %f %float_52
               OpBranch %10
         %10 = OpLabel
               OpLoopMerge %12 %13 None
               OpBranch %11
         %11 = OpLabel
         %17 = OpLoad %float %err
         %18 = OpFDiv %float %float_1 %17
         %19 = OpFAdd %float %float_1 %18
               OpStore %next %19
         %23 = OpLoad %float %next
         %24 = OpBitcast %int %23
               OpStore %bits %24
         %25 = OpLoad %int %bits
         %27 = OpBitwiseAnd %int %25 %int_256
         %30 = OpIEqual %bool %27 %int_0
               OpSelectionMerge %32 None
               OpBranchConditional %30 %31 %32
         %31 = OpLabel
         %33 = OpLoad %float %f
         %34 = OpFAdd %float %33 %float_1
               OpStore %f %34
               OpBranch %13
         %32 = OpLabel
         %36 = OpLoad %float %next
         %37 = OpLoad %float %f
         %38 = OpFSub %float %36 %37
               OpStore %err %38
         %39 = OpLoad %float %next
               OpStore %f %39
               OpBranch %13
         %13 = OpLabel
         %40 = OpLoad %float %err
         %42 = OpFOrdGreaterThan %bool %40 %float_9_99999975en006
               OpBranchConditional %42 %10 %12
         %12 = OpLabel
         %45 = OpLoad %float %f
               OpStore %w %45
               OpReturn
               OpFunctionEnd
```

The loop top is at block `%10`, declaring `%12` as the merge block, and `%13` as the continue target.

The 'continue' in GLSL maps to the branch to `%13` from block `%31`.

The OpBranchConditional at block `%13` has the backedge back to `%10` and the exit out to `%12`.

**Comment by dneto0** [comment]
Tint's SPIR-V --> WGSL functionality has gotten far enough that this is what I get from it:

```rust
entry_point compute = main;

var<workgroup> w : f32;

fn main() -> void {
  var f : f32 = 52.0000000;
  var next : f32 = 52.0000000;
  var err : f32 = 52.0000000;
  var bits : i32 = 0;
  f = 52.0000000;
  loop {
    const x_26 : f32 = err;
    next = (1.00000000 + (1.00000000 / x_26));
    const x_29 : f32 = next;
    bits = as<i32>(x_29);
    const x_31 : i32 = bits;
    if (((x_31 & 256) == 0)) {
      continue;
    }
    const x_36 : f32 = next;
    const x_37 : f32 = f;
    err = (x_36 - x_37);
    const x_39 : f32 = next;
    f = x_39;

    continuing {
      const x_40 : f32 = err;
      if ((x_40 > 9.99999975e-06)) {
      } else {
        break;
      }
    }
  }
  const x_42 : f32 = f;
  w = x_42;
  return;
}
```

**Comment by dneto0** [comment]
I wanted to capture @jdashg's observation that it's as if we had an optional extra "while" clause on the end of the continuing clause:

```
   loop {
      if (...) { .... continue; } 
      ...
      continuing {
              ...
      } while(keep_going_expr) 
   }
```

where  the `keep_going_expr` is computed with terms in scope in the loop or continuing clause, and which itself can't escape the loop (either via break or continue).

**Comment by grorg** [comment]
Discussed at the [2020-06-16 meeting](https://docs.google.com/document/d/11RW9hpfzSvZYpLGBm3uhKwoXDvJI42N7jWP_nOB-k2E/).

**Pull request merged**